language/bytecode-verifier/src/borrow_graph.rs

```diff
@@ -0,0 +1,221 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module defines the (acyclic) borrow graph for the type and memory safety analysis.
+use crate::nonce::Nonce;
+use mirai_annotations::checked_verify;
+use std::collections::{BTreeMap, BTreeSet};
+
+/// There are two types of edges in the borrow graph.
+/// A strong edge carries more information than a weak edge.
+#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+enum EdgeType {
+    Strong,
+    Weak,
+}
+
+/// The label on an edge
+pub type Label<T> = Vec<T>;
+
+/// A labeled edge in borrow graph
+#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct Edge<T> {
+    edge_type: EdgeType,
+    label: Label<T>,
+    to: Nonce,
+}
+
+/// A borrow graph is represented as a map from a source nonce to the set of all edges
+/// coming out of it.
+#[derive(Clone, Debug, Default, Eq, Ord, PartialEq, PartialOrd)]
+pub struct BorrowGraph<T>(BTreeMap<Nonce, BTreeSet<Edge<T>>>);
```

Comment:
Did you consider using an IndexSet?

Reply:
Discussed offline.  

IndexSet and BTreeSet are both deterministic and suitable for use in the VM.  The use of one or the other might be a matter of performance.  We should do a benchmarking study.